---
title: "Kevin_R_Assignment"
author: "Kevin Falk"
date: "October 17, 2017"
output:
  html_document: default
  pdf_document: default
---

####################### Part 1 ##################################
#Load libraries.

```{r}
library(plyr)
library(dplyr)
library(ggplot2)
library(reshape2)
```

#set working directory

**#Make sure you whoever is using sets proper working directory.**

**#I changed setwd to work on my personal directory after I cloned. Will have to set back to your own personal working directory.**

```{r}
setwd("c:/Users/Stock/Desktop/BCB-546X/Kevin_R_Assignment/BCB546X-Fall2017/R_Assignment")
```

#Read in data

**#Not sure why I had to but I had to manually copy and paste the read into R console for it to read in the properly.**

#read in fang_genotypes text file and set it as genotypes

```{r}
library(readr)
genotypes <- read.table("fang_et_al_genotypes.txt", header = T, stringsAsFactors = T)
```

#take a look at the data

```{r}
head(genotypes)
str(genotypes)
nrow(genotypes)
ncol(genotypes)
geno_table <- tbl_df(genotypes)
head(geno_table)
glimpse(geno_table)
data.frame(head(geno_table))
```

#read in snp_position text file and set it as snps

```{r}
snps <- read.delim("snp_position.txt", header = T, stringsAsFactors = F)
```

#take a look at the data

```{r}
head(snps)
str(genotypes)
nrow(snps)
ncol(snps)
snps_table <- tbl_df(snps)
head(snps_table)
glimpse(snps_table)
```

####################################################################
############################# MAIZE ################################

#use the filter verb to screen for just the maize SNPs

```{r}
maize <- filter(genotypes, Group %in% c("ZMMIL","ZMMLR","ZMMMR"))
```

#lets take a look at our new df

```{r}
nrow(maize)
head(maize)
```

#transpose maize df so that it reads horizontally

```{r}
maize_t <- t(maize)
```

#use merge function to combine both files together using the SNP_ID column
#in the snps_table and the dataframe's actual row names in the maize_t df.

```{r}
maize_m <- merge(snps_table, maize_t, by.x = "SNP_ID", by.y = "row.names")
```

#cut out unwanted columns

```{r}
maize_c <- maize_m[,-c(2,5:15)]
```

**#JS EDIT - Instead of just changing the ?/? to -/-, I was under the impression of changing to the "?" symbol or "-" symbol alone. **

#change the ?/? to ? using the apply function

```{r}
df1 <- apply(maize_c,2, function(x) gsub("\\?/\\?", "\\?", x))
```

#change the ?/? to - using the apply funciton

```{r}
df2 <- apply(maize_c,2, function(x) gsub("\\?/\\?", "\\-", x))
```

#apply outputs a matrix, so change it to a dataframe

```{r}
genotypes2 <- as.data.frame(df1)
genotypes3 <- as.data.frame(df2)
```

#reorder dataframe by increasing position number

**#multiple and NA's for position are converted to NA based on code below.**

```{r}
maize_increasing_snps <- genotypes2[order(as.numeric(as.character(genotypes2$Position))),]
```

#reorder dataframe by decreasing position number

```{r}
maize_decreasing_snps <- genotypes3[order(-as.numeric(as.character(genotypes3$Position))),]
```

#Create CSV files

#instead of manually creating each dataframe, we can use a for loop

**#I had to changed the path location for the files as before to get it to work for me.**

```{r}
for (i in 1:10) {
  x <- maize_increasing_snps[maize_increasing_snps$Chromosome == i,]
#we can create a new CSV file for each chromosome
  write.csv(x, sprintf("c:/Users/Stock/Desktop/BCB-546X/Kevin_R_Assignment/BCB546X-Fall2017/R_Assignment/maize_increasing_snps_chromosome_%d.csv", i), row.names = F)
}
```

#instead of manually creating each dataframe, we can use a for loop

```{r}
for (i in 1:10) {
  x <- maize_decreasing_snps[maize_decreasing_snps$Chromosome == i,]
  #we can create a new CSV file for each chromosome
  write.csv(x, sprintf("c:/Users/Stock/Desktop/BCB-546X/Kevin_R_Assignment/BCB546X-Fall2017/R_Assignment/maize_decreasing_snps_chromosome_%d.csv", i), row.names = F)
}
```

####################################################################
########################### TEOSINTE ###############################

#use the filter verb to screen for just the teosinte SNPs

```{r}
teosinte <- filter(genotypes, Group %in% c("ZMPBA","ZMPIL","ZMPJA"))
```

#lets take a look at our new df

```{r}
head(teosinte)
nrow(teosinte)
ncol(teosinte)
```

#transpose teosinte df so that it reads horizontally

```{r}
teosinte_t <- t(teosinte)
```

#use merge function to combine both files together using the SNP_ID column
#in the snps_table and the dataframe's actual row names in the teosinte_t df.

```{r}
teosinte_m <- merge(snps_table, teosinte_t, by.x = "SNP_ID", by.y = "row.names")
```

#cut out unwanted columns

```{r}
teosinte_c <- teosinte_m[,-c(2,5:15)]
```

#change the ?/? to "?" using the apply function
**#JS EDIT - Again, instead of just changing the ?/? to -/-, I was under the impression of changing to the "?" symbol or "-" symbol alone. **

```{r}
df1 <- apply(teosinte_c,2, function(x) gsub("\\?/\\?", "\\?", x))
```

#change the ?/? to "-" using the apply funciton

```{r}
df2 <- apply(teosinte_c,2, function(x) gsub("\\?/\\?", "\\-", x))
```

#apply outputs a matrix, so change it to a dataframe

```{r}
genotypes2 <- as.data.frame(df1)
genotypes3 <- as.data.frame(df2)
```

#reorder dataframe by increasing position number

```{r}
teosinte_increasing_snps <- genotypes2[order(as.numeric(as.character(genotypes2$Position))),]
```

#reorder dataframe by decreasing position number

```{r}
teosinte_decreasing_snps <- genotypes3[order(-as.numeric(as.character(genotypes3$Position))),]
```

#Create CSV Teosinte files

#instead of manually creating each dataframe, we can use a for loop
**#Had to manually add in file location to get it to save in my system.**

```{r}
for (i in 1:10) {
  x <- teosinte_increasing_snps[teosinte_increasing_snps$Chromosome == i,]
  #we can create a new CSV file for each chromosome
  write.csv(x, sprintf("c:/Users/Stock/Desktop/BCB-546X/Kevin_R_Assignment/BCB546X-Fall2017/R_Assignment/teosinte_increasing_snps_chromosome_%d.csv", i), row.names = F)
}
```

#instead of manually creating each dataframe, we can use a for loop

```{r}
for (i in 1:10) {
  x <- teosinte_decreasing_snps[teosinte_decreasing_snps$Chromosome == i,]
  #we can create a new CSV file for each chromosome
  write.csv(x, sprintf("c:/Users/Stock/Desktop/BCB-546X/Kevin_R_Assignment/BCB546X-Fall2017/R_Assignment/teosinte_decreasing_snps_chromosome_%d.csv", i), row.names = F)
}
```

####################### Part 2 ##################################

#transpose genotypes df so that it reads horizontally

```{r}
genotypes_t <- t(genotypes)
```

#use merge function to combine both files together using the SNP_ID column
#in the snps_table and the dataframe's actual row names in the maize_t df.

```{r}
genotypes_m <- merge(snps, genotypes_t, by.x = "SNP_ID", by.y = "row.names", all.y = TRUE)
```

#Pull the name of all the SNPs from the original file

```{r}
SNP_names <- colnames(genotypes)[-c(1:3)]
```

#reshape the data using the melt command

```{r}
alleles <- melt(genotypes, measure.vars = SNP_names)
```

#look at the just the OTU and Groups and SNPs

```{r}
genotypes_n <- genotypes_m[,-c(1:15)]
```

#Make the OTU Groups the header

```{r}
colnames(genotypes_n) <- as.character(unlist(genotypes_n[14,]))
```

#Remove the OTU and Groups rows

```{r}
genotypes_n <- genotypes_n[-c(14,16),]
View(genotypes_n)
```

#look at the just the other stuff

```{r}
genotypes_o <- genotypes_m[,c(1:15)]
```

#remove the rows of OTU and Groups

```{r}
genotypes_o <- genotypes_o[-c(14,16),]
```

#bind both dataframes together in new format

```{r}
genotypes_p <- cbind(genotypes_o, genotypes_n)
genotypes.p <- melt(genotypes_p)
```

#reorder dataframe by increasing position number

```{r}
genotypes_s <- genotypes_m[order(as.numeric(as.character(genotypes_m$Chromosome))),]

genotypes_s <- tbl_df(genotypes_s)
```

#plot the SNPs per chromosome

```{r}
ggplot(genotypes_s, aes(x=Chromosome)) + geom_bar()
```

#Plot the SNPs per chromosome

#create a summary table of SNPs per chromosome
**#Original code threw error, this was suggested fix.**

```{r}
detach("package:plyr", unload=TRUE)
library(dplyr)
x <- genotypes_s %>% group_by(Chromosome) %>% summarise(SNP_count = n())
```

#turn columns into factors so that we can plot them

```{r}
x$Chromosome <- as.factor(x$Chromosome)
x$SNP_count <- as.factor(x$SNP_count)
```

#plot the SNPs per chromosome

```{r}
plot(x$Chromosome, x$SNP_count, xlab = "Chromosome", ylab = "SNP Count")
```

#in general, there is a decreasing number of SNPs from the first to tenth chromosome

#Missing data and amount of heterozygosity

#Create a new column to indicate whether a particular site is homozygous

```{r}
alleles <- apply(alleles,2, function(x) gsub("C/C", TRUE, x))
alleles <- apply(alleles,2, function(x) gsub("T/T", TRUE, x))
alleles <- apply(alleles,2, function(x) gsub("A/A", TRUE, x))
alleles <- apply(alleles,2, function(x) gsub("G/G", TRUE, x))
alleles <- apply(alleles,2, function(x) gsub("G/T", FALSE, x))
alleles <- apply(alleles,2, function(x) gsub("A/C", FALSE, x))
alleles <- apply(alleles,2, function(x) gsub("A/G", FALSE, x))
alleles <- apply(alleles,2, function(x) gsub("A/T", FALSE, x))
alleles <- apply(alleles,2, function(x) gsub("C/G", FALSE, x))
alleles <- apply(alleles,2, function(x) gsub("C/T", FALSE, x))
alleles <- apply(alleles,2, function(x) gsub("C/A", FALSE, x))
alleles <- apply(alleles,2, function(x) gsub("G/A", FALSE, x))
alleles <- apply(alleles,2, function(x) gsub("G/C", FALSE, x))
alleles <- apply(alleles,2, function(x) gsub("T/A", FALSE, x))
alleles <- apply(alleles,2, function(x) gsub("T/C", FALSE, x))
alleles <- apply(alleles,2, function(x) gsub("T/G", FALSE, x))  
```

#Recode the missing data as NA

```{r}
alleles <- apply(alleles,2, function(x) gsub("\\?/\\?", NA, x))  
```

#Data Visualization

#Name columns in the data set

```{r}
colnames(alleles)[1:5] <- c("Sample_ID","JG_OTU","Group","SNP_ID","Homozygous")
alleles_df <- as.data.frame(alleles)
```

#Sort your dataframe using Group and Species_ID values

```{r}
alleles_s <-arrange(alleles_df, Sample_ID, Group)
```

#Count the number of Homozygous (TRUE) and Heterozygous (FALSE)

```{r}
alleles_tf <- alleles_s %>% 
  group_by(Homozygous) %>% 
  tally(sort=TRUE)
```

#Use ggplot to plot the number of Homozygous, Non-homozygous and NA samples

```{r}
ggplot(data = alleles_tf, aes(x = Homozygous, y = as.numeric(n))) + geom_bar(stat = "identity", position = "stack")
```

#create dataframe that compares Group with Homozygousity

```{r}
group_homo <- alleles_s %>%
  group_by(Group) %>%
  select(Homozygous) %>%
  data.frame()
```

#Visualize one other feature of the dataset.
#Histogram comparing homozygous loci and species grouping

```{r}
ggplot(data = group_homo, aes(x = Group, y = Homozygous))+ geom_bar(stat = "identity", position = "stack")
```


**#Joe Stock Edits**
**#Mostly just cosmetic. Nice code all the way around. I cleaned up the R code so that you can run it within the chunks. Use of the Insert R tab above makes things easier to view and run. For whatever reason I could not get my working directory to funciton properly so I had to input the entire path to whatever file I was trying to load or save. Nice Work!!!**